### PR TITLE
adds proposed spec for cache

### PR DESCRIPTION
**Pull request by dmfenton**
@chelm @ngoldman @jgravois I could use some review here.

### CODE REVIEW

**Comment by dmfenton** [comment]
I think for the next version of Koop, we should standardize the way table is passed to the cache. Right now it's a mix of key and table + layer.

File: doc/cache-spec.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+* Layer (integer): the id of the layer. Usually defaults to 0
+* Callback (function): returns err on failure or true on success
+
+### remove
+```javascript
+function (table, function(err, success))
+```
+* Table: the name of the table to be deleted. Includes the layer id 
+ * e.g. Socrata:ejk5-56ko:0
+* Callback: returns err on failure or true on success
```
Comment (jgravois) [inline doc/cache-spec.md]: just a gut reaction, but it feels weird to me that the callback returns either an `err` or a `boolean`

File: doc/cache-spec.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+* Table (string): the name of the table to be inserted. Usually of the form: type:key
+* Data (object): geojson to be inserted after creation
+* Layer (integer): the id of the layer. Usually defaults to 0
+* Callback (function): returns err on failure or true on success
+
+### remove
+```javascript
+function (table, function(err, success))
+```
+* Table: the name of the table to be deleted. Includes the layer id
```
Comment (jgravois) [inline doc/cache-spec.md]: you use `layerId` everywhere else

File: doc/cache-spec.md (hunk 1 above)
Comment (ngoldman) [inline doc/cache-spec.md]: The standard way to deal with callbacks in node is `function (err, params*)`, where `err` is either an [error object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) or `null`. If there's an error nothing else should be passed. If `err` is `null` then response data should be passed afterwards (there's no limit to number of params but it's usually best to limit it to one unless really necessary).

This article gives a good overview of the convention: http://thenodeway.io/posts/understanding-error-first-callbacks/

Comment (jgravois) [inline doc/cache-spec.md]: thanks for that.  i'm familiar with the node convention of passing back `errors

Comment (ngoldman) [inline doc/cache-spec.md]: ah ok. I've seen returning true on success in a lot of other places.. the mix with error first callbacks is a little weird. it's late on the east coast but i think i remember seeing some programs taking callback(null) as the indicator of success but don't quote me on it. reasonable alternatives welcome

Comment (chelm) [inline doc/cache-spec.md]: Yeah this is worth reviewing and cleaning up. Most of the time the `success` is pretty much ignored and the existence of an `error` in the callback is used. I think we can remove `success` in most cases

File: doc/cache-spec.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
+# Spec for Koop cache modules
+In order to function as a cache for Koop, a module must support the following API
+
+## Functions
+
+### insert
```
Comment (ajturner) [inline doc/cache-spec.md]: how about `createTable` for a new table and `insert` adds rows similar to sql

**Comment by chelm** [comment]
What I really like about this is it will force us to closely look at what we pass into various methods. It a great opp to address consistency between the cache methods, koop's Cache wrapper and each provider.

@dmfenton is this doc missing the `services*` methods?

**Comment by dmfenton** [comment]
That was definitely my goal here. Let's get existing doc'd then we can propose a better version for the refactor.

I don't see references to any services methods in koop-pgcache or lib/cache. What are you referring to?

**Comment by chelm** [comment]
@dmfenton - maybe it just not meant to be complete? but in pgcache there is https://github.com/Esri/koop-pgcache/blob/master/index.js#L542-L601

I think the problem is that the providers that use those methods must not be going through lib/Cache which needs to change.

**Comment by dmfenton** [comment]
Oh, I see. Yeah, I'm not sure that is happening in lib/Cache.

**Comment by ngoldman** [comment]
@dmfenton merging for now as it's a good working draft, we can continue the discussion and revise as necessary

**Pull request merged**